internal/operands/common-instancetypes/reconcile.go

```diff
@@ -58,52 +98,196 @@ func fetchClusterResources[C clusterType](bundlePath string) ([]C, error) {
 }
 
-func (c *commonInstancetypes) WatchTypes() []operands.WatchType {
+func reconcileRemovedInstancetypes(request *common.Request, existingResources []instancetypev1alpha2.VirtualMachineClusterInstancetype, resourcesFromURL []instancetypev1alpha2.VirtualMachineClusterInstancetype) error {
+	resourceFromURLByName := make(map[string]instancetypev1alpha2.VirtualMachineClusterInstancetype)
+	for _, resourceFromURL := range resourcesFromURL {
+		resourceFromURLByName[resourceFromURL.Name] = resourceFromURL
+	}
+	for _, existingResource := range existingResources {
+		if _, resourceProvided := resourceFromURLByName[existingResource.Name]; !resourceProvided {
+			request.Logger.Info(fmt.Sprintf("removing the no longer provided %s VirtualMachineClusterInstancetype", existingResource.Name))
+			if err := request.Client.Delete(request.Context, &existingResource); err != nil {
+				return err
+			}
+		}
+	}
 	return nil
 }
 
-func (c *commonInstancetypes) RequiredCrds() []string {
-	return []string{
-		virtualMachineClusterInstancetypeCrd,
-		virtualMachineClusterPreferenceCrd,
+func reconcileRemovedPreferences(request *common.Request, existingResources []instancetypev1alpha2.VirtualMachineClusterPreference, resourcesFromURL []instancetypev1alpha2.VirtualMachineClusterPreference) error {
+	resourceFromURLByName := make(map[string]instancetypev1alpha2.VirtualMachineClusterPreference)
+	for _, resourceFromURL := range resourcesFromURL {
+		resourceFromURLByName[resourceFromURL.Name] = resourceFromURL
+	}
+	for _, existingResource := range existingResources {
+		if _, resourceProvided := resourceFromURLByName[existingResource.Name]; !resourceProvided {
+			request.Logger.Info(fmt.Sprintf("removing the no longer provided %s VirtualMachineClusterPreference", existingResource.Name))
+			if err := request.Client.Delete(request.Context, &existingResource); err != nil {
+				return err
+			}
+		}
 	}
+	return nil
 }
 
-func (c *commonInstancetypes) Reconcile(request *common.Request) ([]common.ReconcileResult, error) {
-	request.Logger.Info("Reconciling common-instancetypes")
+func (c *commonInstancetypes) reconcileFromURL(request *common.Request) ([]common.ReconcileResult, error) {
+	// TODO - In the future we should handle cases where the URL remains the same but the provided resources change.
+	if c.resourceURL != "" && c.resourceURL == *request.Instance.Spec.CommonInstancetypes.URL {
+		request.Logger.Info(fmt.Sprintf("Skipping reconcile of common-instancetypes from URL %s, force with a restart of the service.", *request.Instance.Spec.CommonInstancetypes.URL))
+		return nil, nil
+	}
+
+	// Stash the URL
+	c.resourceURL = *request.Instance.Spec.CommonInstancetypes.URL
+
+	request.Logger.Info(fmt.Sprintf("Reconciling common-instancetypes from URL %s", *request.Instance.Spec.CommonInstancetypes.URL))
+	clusterInstancetypesFromURL, clusterPreferencesFromURL, err := FetchResourcesFromURL(*request.Instance.Spec.CommonInstancetypes.URL)
```

Comment:
This line fetches the contents of the URL on every reconcile cycle. Do we want to support reloading the contents of the URL if they change? If not, can you cache the resources, so the URL is accessed only when the `.spec.commonInstancetypes.URL` changes?

Reply:
See the conditional above, it's only triggered when the operand is first called or when the URL changes at the moment, we skip the rest of the time given the constraints that the URL needs to target a fixed `?ref=` or `?target=`.

Reply:
Ok, I missed that condition.